contracts/lib/TimeLockUpgradeV2.sol

```diff
@@ -0,0 +1,152 @@
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+pragma solidity 0.5.7;
+
+import { Ownable } from "openzeppelin-solidity/contracts/ownership/Ownable.sol";
+import { SafeMath } from "openzeppelin-solidity/contracts/math/SafeMath.sol";
+
+
+/**
+ * @title TimeLockUpgradeV2
+ * @author Set Protocol
+ *
+ * The TimeLockUpgradeV2 contract contains a modifier for handling minimum time period updates
+ *
+ * CHANGELOG:
+ * - Requires that the caller is the owner
+ * - New function to allow deletion of existing timelocks
+ */
+contract TimeLockUpgradeV2 is
+    Ownable
+{
+    using SafeMath for uint256;
+
+    /* ============ State Variables ============ */
+
+    // Timelock Upgrade Period in seconds
+    uint256 public timeLockPeriod;
+
+    // Mapping of upgradable units and initialized timelock
+    mapping(bytes32 => uint256) public timeLockedUpgrades;
+
+    /* ============ Events ============ */
+
+    event UpgradeRegistered(
+        bytes32 _upgradeHash,
+        uint256 _timestamp
+    );
+
+    event UpgradeUnregistered(
+        bytes32 _upgradeHash
+    );
+
+    /* ============ Modifiers ============ */
+
+    modifier timeLockUpgrade() {
+        require(
+            isOwner(),
+            "timeLockUpgradeV2: The caller must be the owner"
+        );
+
+        // If the time lock period is 0, then allow non-timebound upgrades.
+        // This is useful for initialization of the protocol and for testing.
+        if (timeLockPeriod == 0) {
+            _;
+
+            return;
+        }
+
+        // The upgrade hash is defined by the hash of the transaction call data,
+        // which uniquely identifies the function as well as the passed in arguments.
+        bytes32 upgradeHash = keccak256(
+            abi.encodePacked(
+                msg.data
+            )
+        );
+
+        uint256 registrationTime = timeLockedUpgrades[upgradeHash];
+
+        // If the upgrade hasn't been registered, register with the current time.
+        if (registrationTime == 0) {
+            timeLockedUpgrades[upgradeHash] = block.timestamp;
+
+            emit UpgradeRegistered(
+                upgradeHash,
+                block.timestamp
+            );
+
+            return;
+        }
+
+        require(
+            block.timestamp >= registrationTime.add(timeLockPeriod),
+            "TimeLockUpgradeV2: Time lock period must have elapsed."
+        );
+
+        // Reset the timestamp to 0
+        timeLockedUpgrades[upgradeHash] = 0;
+
+        // Run the rest of the upgrades
+        _;
+    }
+
+    /* ============ Function ============ */
+
+    /**
+     * Removes an existing upgrade.
+     *
+     * @param  _upgradeHash    Keccack256 hash that uniquely identifies function called and arguments 
+     */
+    function deregisterUpgrade(
```

Comment:
We should settle on unregister (used in event) vs deregister (used here). Would also be fine with `removeRegisteredUpgrade` or `removeUpgrade` as function name and UpgradeRemoved as event. Deregister and Unregister are weird words. Either way...consistency.